### PR TITLE
ModuleManifest::getModuleByPath fix to ensure right module is returned

### PR DESCRIPTION
**Pull request by sunnysideup**
see: https://github.com/silverstripe/silverstripe-framework/issues/9561 for details.

I have added forward slash for more reliable matching AND
I have added a unit test that passes AFAIK.

### CODE REVIEW

File: tests/php/Core/Manifest/ModuleManifestTest.php [hunk 1]
```diff
@@ -105,4 +107,35 @@ public function testGetResourcePathOnRoot()
...
         );
     }
+
+    /**
+     * @return array
+     */
+    public function providerTestGetModuleByPath()
+    {
+        return [
+            // ['vendor/silverstripe/module/classes/classA.php', 'module'],
```
Comment (emteknetnz) [inline tests/php/Core/Manifest/ModuleManifestTest.php:117]: Is this line meant to be commented out?

File: tests/php/Core/Manifest/fixtures/classmanifest/vendor/silverstripe/modulecbetter/_config/config.yml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
```
Comment (emteknetnz) [inline tests/php/Core/Manifest/fixtures/classmanifest/vendor/silverstripe/modulecbetter/_config/config.yml:1]: Are these various `/Manifest/` files supposed to be in the PR?  They seem odd

**Comment by emteknetnz** [comment]
@sunnysideup would you be able to update the title on this PR to something a little more descriptive, cheers

File: tests/php/Core/Manifest/ModuleManifestTest.php (hunk 1 above)
Comment (sunnysideup) [reply tests/php/Core/Manifest/ModuleManifestTest.php:117]: Yes, I will remove it.

File: tests/php/Core/Manifest/fixtures/classmanifest/vendor/silverstripe/modulecbetter/_config/config.yml (hunk 2 above)
Comment (sunnysideup) [reply tests/php/Core/Manifest/fixtures/classmanifest/vendor/silverstripe/modulecbetter/_config/config.yml:1]: Yes, we basically create a new module with some basic stuff

**Comment by sunnysideup** [comment]
Do we need to do anything else?

**Comment by sunnysideup** [comment]
> @sunnysideup would you be able to update the title on this PR to something a little more descriptive, cheers

Done ... not sure if it is any better as I am not sure what is customary.

**Comment by robbieaverill** [comment]
Hi @sunnysideup, it looks like this PR has broken a bunch of our unit tests. Would you mind fixing those up?

**Comment by sunnysideup** [comment]
Yeah - that does not look good! @robbieaverill  - will look at this!

**Comment by sunnysideup** [comment]
I fixed my specific tests, but I am not sure why other tests are broken.  Is that normal?

**Comment by robbieaverill** [comment]
I would suggest that the changes you've made are causing those test failures, particularly the ones that are saying it can't resolve the silverstripe/admin module

**Comment by sunnysideup** [comment]
Hmmm, yes. But some of the errors are really confusing to me.

**Comment by sunnysideup** [comment]
I am getting different errors locally.

**Comment by sunnysideup** [comment]
adding this test:

```php
    /**
     * @return array
     */
    public function providerTestGetModuleByPath()
    {
        return [
            ['vendor/silverstripe/modulec/code/VendorClassA.php', 'silverstripe/modulec'],
            ['vendor/silverstripe/modulecbetter/code/VendorClassX.php', 'silverstripe/modulecbetter'],
        ];
    }
    
    /**
     * @dataProvider providerTestGetModuleByPath
     * @param string $path
     * @param string $expectedModuleName
     */
    public function testGetModuleByPath($path, $expectedModuleName)
    {
        // important - load the manifest that we are working with to the ModuleLoader
        ModuleLoader::inst()->pushManifest($this->manifest);
    
        // work out variables
        $path = $this->base . '/' . $path;
        $module = $this->manifest->getModuleByPath($path);
        $actualModuleName = $module->getName();
    
        // it is testing time!
        $this->assertEquals($expectedModuleName, $actualModuleName);
    }
```

I lack the knowledge on how to fix this and it appears to me that the test suite is broken rather than my code.


breaks a whole bunch of other tests ....

**Comment by sunnysideup** [comment]
I think I cracked it!

**Comment by sunnysideup** [comment]
@robbieaverill - is this better?

**Comment by sunnysideup** [comment]
@robbieaverill I have linted to code and that now seems to pass.

I would love to run my linter on all of framework.  Could I do that as a pull request?

**Comment by ScopeyNZ** [comment]
Probably not. Your linting includes rules like putting a space after `!` in conditionals, which we don't have a specific rule for (apparently), but we do have a de-facto standard to _not_ include that space.

A linting PR like that would just enforces a bunch of arbitrary new rules that we would need to discuss.

**Comment by robbieaverill** [comment]
@sunnysideup I only work part time on the Silverstripe project nowadays, apologies if I don't get to this promptly.

**Comment by sunnysideup** [comment]
@ScopeyNZ - no worries. Fixed. Your lint test should pick that up. 

I have removed the exclamation mark space.  What linter do you use?  I use this one:

- easy coding standards (https://packagist.org/packages/symplify/easy-coding-standard) with this config: 

- https://github.com/sunnysideup/silverstripe-easy-coding-standards/blob/master/ecs.yml

**Comment by sunnysideup** [comment]
> @sunnysideup I only work part time on the Silverstripe project nowadays, apologies if I don't get to this promptly.

All good. Sorry for all the noise!

**Comment by sunnysideup** [comment]
> Probably not. Your linting includes rules like putting a space after `!` in conditionals, which we don't have a specific rule for (apparently), but we do have a de-facto standard to _not_ include that space.
> 
> A linting PR like that would just enforces a bunch of arbitrary new rules that we would need to discuss.

Obviously there would be a super clear discussion about the rule-set applied.  Probably a lot of hassle to even get done, but then it would be good to always apply that ruleset automatically at every commit or something like that - so that even sloppy coders just end up with the same clean code. That could save a lot of time.

**Comment by robbieaverill** [comment]
Hey @sunnysideup, our linting rules are PSR-12. We are aiming for full compatibility in SilverStripe 5.x. You can see the rules per module in `phpcs.xml` for each module. These are configuration for PHP CodeSniffer (the linting tool). We have some opinionated rules on top of PSR-12, but not many. If you'd like to discuss our linting rules further I'd suggest opening an RFC on framework with your recommended changes 😄

**Comment by sunnysideup** [comment]
> Hey @sunnysideup, our linting rules are PSR-12. We are aiming for full compatibility in SilverStripe 5.x. You can see the rules per module in `phpcs.xml` for each module. These are configuration for PHP CodeSniffer (the linting tool). We have some opinionated rules on top of PSR-12, but not many. If you'd like to discuss our linting rules further I'd suggest opening an RFC on framework with your recommended changes

Thank you for explaining that. 

In terms of changing / discussing: yeah nah.... the less I have to think about that the better.

**Comment by sunnysideup** [comment]
@robbieaverill @emteknetnz - what do we need to do to get this released?

**Comment by kinglozzer** [comment]
Thanks for your work on this so far @sunnysideup. I think the logic in the `getModuleByPath()` method is a little more verbose than it needs to be - I appreciate you’re probably trying to retain the existing `strpos()` logic, but I think if we trim the directory separators we can just remove that and compare the two strings directly:

```php
public function getModuleByPath($path)
{
    // Ensure path exists
    $path = trim(realpath($path), DIRECTORY_SEPARATOR);
    if (!$path) {
        return null;
    }

    /** @var Module $rootModule */
    $rootModule = null;

    // Find based on loaded modules
    $modules = ModuleLoader::inst()->getManifest()->getModules();

    foreach ($modules as $module) {
        // If the module has a path, check whether it matches - otherwise move on to next module
        $modulePath = trim(realpath($module->getPath()), DIRECTORY_SEPARATOR);
        if ($modulePath && $modulePath !== $path) {
            continue;

        }
        
        // If this is the root module, keep looking in case there is a more specific module later
        if (empty($module->getRelativePath())) {
            $rootModule = $module;
        } else {
            return $module;
        }
    }

    // Fall back to top level module
    return $rootModule;
}
```

**Comment by sunnysideup** [comment]
Sure, but I think the real issue is the reverse logic... We should say, instead: 

```
if modulePath is the same as the start of the Path then return the ModulePath
```

But does it really matters if it works? I'd rather just have a it fixed now.

**Comment by sunnysideup** [comment]
@kinglozzer 
what are we going to do here?

**Comment by kinglozzer** [comment]
Sorry for the delay! I was completely wrong with my suggestion above anyway, I misunderstood how the method works 🤦‍♂️ 

I do still think the logic in `getModuleByPath()` should be simplified though (sorry...), I had to re-read it a few times to work out why we're removing and then re-adding the directory separators. How about this:

```php
foreach ($modules as $module) {
    $modulePath = realpath($module->getPath());
    // If this module has a path
    if ($modulePath) {
        // Force a trailing slash so we don't mix up, for example:
        // silverstripe/framework and silverstripe/framework-extension
        $modulePath = $modulePath . DIRECTORY_SEPARATOR;

        // If the path doesn't begin with this module's path, skip this module
        if (stripos($path, $modulePath) !== 0) {
            continue;
        }
    }

    // ...
}
```

**Comment by sunnysideup** [comment]
@kinglozzer - that is better, BUT ... ;-) sometimes the `path` is the `$modulePath` without the final `DIRECTORY_SEPARATOR` - so we need to check for both ;-)

**Comment by kinglozzer** [comment]
> that is better, BUT ... ;-) sometimes the path is the $modulePath without the final DIRECTORY_SEPARATOR - so we need to check for both.

Gah, I was going by the docblock which says `$path` is a path to a specific file!

Let’s just get this bug fixed. The extra unit test coverage will make refactoring it in future easier anyway if anyone decides to take that on. Sorry for the noise @sunnysideup 😅 

Could you squash the changes into a single commit?

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**